### ISSUE TITLE
Stream instead of buffer results

### ISSUE THREAD

**Issue by TheSlimvReal** [Aam-Digital/replication-backend#109]
Normal `async`/`await` NodeJS fetch requires to load all data of a request and buffer it.
Using NodeJS streams its possible to read the request line by line and directly forward the results. 
The reduces the amount of data that needs to be buffered.
The difficulty is that the results cannot be parsed as plain JSON but need to be adjusted.

**Comment by sleidig** [comment]
## Implementation plan

### Why this matters

Every proxied endpoint today follows this pipeline: axios buffers the **entire** CouchDB response (and decompresses it) → `JSON.parse` builds a full object graph (~2–3× raw size) → filtering creates a partial copy → Nest `JSON.stringify`s the whole result again. Peak memory is roughly **4–5× the raw response size**, multiplied by concurrent syncing clients. `GET /:db/_all_docs?include_docs=true` without a limit can pull the entire database into heap in a single request.

### Scope

Stream **CouchDB → client responses** for the read-heavy endpoints, in order of memory impact:

| Phase | Endpoint | Response shape to stream |
|---|---|---|
| 2 | `GET/POST /:db/_all_docs` | `{total_rows, offset, rows: [...]}` |
| 3 | `POST /:db/_bulk_get` | `{results: [...]}` |
| 4 | `GET /:db/_changes` | `{results: [...], last_seq, pending, lostPermissions}` |
| 5 | `POST /:db/_find` | `{docs: [...], bookmark, warning}` |

Request streaming (`_bulk_docs` uploads) stays **out of scope** — bodies are already capped at 10 MB by the JSON body-parser limit, and the permission check needs the full doc list anyway.

### Phase 1 — streaming infrastructure

1. `CouchdbService.getStream()` / `postStream()` returning the raw axios stream (`responseType: 'stream'`), sending `Accept-Encoding: identity` on these internal requests. *Gotcha: axios does not auto-decompress stream responses — disabling gzip on the (local, fast) backend↔CouchDB hop avoids a gunzip step; client-facing compression comes from the `compression` middleware instead (see #261).*
2. Add **`stream-json`** as the incremental parser. Recommended over the "read line by line" idea from the issue description: CouchDB does emit one row per line for `_all_docs`/`_changes`, but that is an undocumented formatting detail — `stream-json` parses actual JSON tokens and won't break across CouchDB versions or intermediate proxies.
3. One reusable utility, e.g. `JsonArrayFilterStream`:
   - Input: upstream Readable, the path of the array to stream (`rows` / `results` / `docs`), a per-item `filter(item) => boolean | transformed item`, pass-through for the non-array envelope fields (`total_rows`, `last_seq`, …) re-emitted in place.
   - Output: valid JSON written incrementally to the Express `Response` via `stream.pipeline(...)` (backpressure + cleanup on client disconnect).
   - **Error semantics:** upstream failure *before* the first byte → normal HTTP error response. Failure *mid-stream* → headers already sent, so `res.destroy()` to force a truncated response (PouchDB treats it as a network error and retries); log + report to Sentry. Must be covered by tests.
4. Unit-test the transform hard: filtered items, empty arrays, envelope fields before *and* after the array, malformed upstream JSON, client abort.

### Phase 2 — `_all_docs`

Controllers switch to `@Res()` + pipeline; the per-row predicate is exactly the current logic from `BulkDocumentService.filterAllDocsResponse` (`isReplicable` + `_deleted || ability.can('read', row.doc)`). Keep the predicate in the service so the logic stays unit-testable without streams.

### Phase 3 — `_bulk_get`

Per-`result` filtering of nested `docs` (logic from `filterBulkGetResponse`), dropping results whose `docs` end up empty. Items are self-contained, so this maps 1:1 onto the utility. Biggest win for initial replication, especially with `attachments=true` where docs carry base64 payloads.

### Phase 4 — `_changes`

Keep the iterative fetch loop and the 8 s time budget (they exist for good reasons — browser idle timeouts), but write each iteration's permitted results to the response immediately instead of accumulating: emit `{"results":[`, stream items (omitting `doc` per item unless `include_docs=true`), then close with `],"last_seq":"...","pending":N,"lostPermissions":[...]}` after the loop.

Bonus: first bytes flow early, reducing the browser idle-timeout pressure that motivated `MAX_PROCESSING_TIME_MS` (keep the budget anyway). The internal per-iteration CouchDB fetches can stay buffered initially (bounded at 1000 docs) and be converted to streaming later if profiling says so.

### Phase 5 — `_find`

Same utility, predicate from `filterFindResponse`.

### Phase 6 — supporting changes

- `compression()` middleware for client-facing responses (verify it flushes on streamed chunks) — overlaps with #261.
- A load-test script (seed a CouchDB with ~100k docs in docker, replicate with a permission-restricted user, record RSS before/after) to prove the memory ceiling and catch regressions.

### Risks / trade-offs

- **Status codes:** streaming commits to `200` before the body is complete — mitigated by the truncate-on-error semantics above.
- **`lostPermissions` ordering:** the custom field moves to the end of the `_changes` JSON object. Key order doesn't matter to JSON parsers, so clients using `response.json()` are unaffected — verify the ndb-core fetch wrapper does that.
- **e2e tests** must consume SuperTest response streams instead of asserting on parsed objects; budget time for that.
- The non-streaming quick wins are tracked in #261 and should land first — they reduce baseline load while this work proceeds incrementally.
